File: filmy-app/app/src/main/java/rk/entertainment/filmy/modules/movieDetails/CustomPagerAdapter.java
package rk.entertainment.filmy.modules.movieDetails;

import android.content.Context;
import android.graphics.Bitmap;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.PagerAdapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.BitmapTransitionOptions;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;

import java.util.List;

import rk.entertainment.filmy.R;
import rk.entertainment.filmy.models.moviesDetails.BackdropData;
import rk.entertainment.filmy.network.APIUtils;
import rk.entertainment.filmy.utils.GlideApp;

public class CustomPagerAdapter extends PagerAdapter {

    private Context mContext;
    private LayoutInflater mLayoutInflater;
    private List<BackdropData> backdropDataList;

    CustomPagerAdapter(Context context, List<BackdropData> backdropDataList) {
        mContext = context;
        this.backdropDataList = backdropDataList;
        mLayoutInflater = (LayoutInflater) mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    void addAll(List<BackdropData> backdropDataList) {
        this.backdropDataList.addAll(backdropDataList);
        if (this.backdropDataList.size() == 0)
            this.backdropDataList.add(0, new BackdropData());
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return backdropDataList.size();
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == object;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        View itemView = mLayoutInflater.inflate(R.layout.pager_item, container, false);
        ImageView imageView = itemView.findViewById(R.id.iv_pager);
        BackdropData data = backdropDataList.get(position);
        if (data.getFilePath() != null) {
            String backDropUrl = APIUtils.IMAGE_BASE_URL
                    + APIUtils.BACKDROP_IMAGE_SIZE + backdropDataList.get(position).getFilePath();

            Log.i("CustomPagerAdapter ", "instantiateItem backDrop "+backDropUrl);

            GlideApp.with(mContext)
                    .asBitmap()
                    .load(backDropUrl)
                    .placeholder(R.drawable.loading)
                    .transition(BitmapTransitionOptions.withCrossFade())
                    .listener(new RequestListener<Bitmap>() {
                        @Override
                        public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Bitmap> target, boolean isFirstResource) {

                            return false;
                        }

                        @Override
                        public boolean onResourceReady(Bitmap resource, Object model, Target<Bitmap> target, DataSource dataSource, boolean isFirstResource) {
                            imageView.setScaleType(ImageView.ScaleType.FIT_XY);
                            imageView.setImageBitmap(resource);
                            return false;
                        }
                    })
                    .into(imageView);
        } else {
            imageView.setScaleType(ImageView.ScaleType.FIT_CENTER);
            imageView.setImageResource(R.drawable.loading);
        }
        container.addView(itemView);
        return itemView;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        container.removeView((LinearLayout) object);
    }
}

File: filmy-app/app/src/main/java/rk/entertainment/filmy/view/features/movieDetails/CustomPagerAdapter.kt
package rk.entertainment.filmy.view.features.movieDetails

import android.content.Context
import android.graphics.Bitmap
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.viewpager.widget.PagerAdapter
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.BitmapTransitionOptions
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import rk.entertainment.filmy.R
import rk.entertainment.filmy.data.models.moviesDetails.BackdropData
import rk.entertainment.filmy.data.repository.APIUtils
import rk.entertainment.filmy.utils.GlideApp
import timber.log.Timber

class CustomPagerAdapter internal constructor(private val mContext: Context, private val backdropDataList: ArrayList<BackdropData>) : PagerAdapter() {

    private val mLayoutInflater: LayoutInflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    fun addAll(backdropDataList: List<BackdropData>) {
        this.backdropDataList.addAll(backdropDataList)
        if (this.backdropDataList.isEmpty())
            this.backdropDataList.add(0, BackdropData())
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return backdropDataList.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val itemView = mLayoutInflater.inflate(R.layout.pager_item, container, false)
        try {
            val imageView = itemView.findViewById<ImageView>(R.id.iv_pager)
            val data = backdropDataList[position]

            if (data.filePath.isNotEmpty()) {

                val backDropUrl = APIUtils.IMAGE_BASE_URL + APIUtils.BACKDROP_IMAGE_SIZE + backdropDataList[position].filePath
                Timber.d("instantiateItem backDrop %s", backDropUrl)

                GlideApp.with(mContext)
                        .asBitmap()
                        .load(backDropUrl)
                        .placeholder(R.drawable.loading)
                        .transition(BitmapTransitionOptions.withCrossFade())
                        .listener(object : RequestListener<Bitmap?> {
                            override fun onLoadFailed(e: GlideException?, model: Any, target: Target<Bitmap?>, isFirstResource: Boolean): Boolean {
                                return false
                            }

                            override fun onResourceReady(resource: Bitmap?, model: Any, target: Target<Bitmap?>, dataSource: DataSource, isFirstResource: Boolean): Boolean {
                                imageView.scaleType = ImageView.ScaleType.FIT_XY
                                imageView.setImageBitmap(resource)
                                return false
                            }
                        })
                        .into(imageView)

            } else {
                imageView.scaleType = ImageView.ScaleType.FIT_CENTER
                imageView.setImageResource(R.drawable.loading)
            }

            container.addView(itemView)
        } catch (e: Exception) {
            Timber.e(e)
        }
        return itemView
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as LinearLayout)
    }
}